### PR TITLE
Reopened/Fix: User experienced big delay when posting the messages

### PR DESCRIPTION
**Pull request by waterim**
### Details

This PR is about to add a skew to Onyx dates for actions
### Fixed Issues

$ https://github.com/Expensify/App/issues/21490
PROPOSAL: $ https://github.com/Expensify/App/issues/21490#issuecomment-1697412148

### Tests

Test steps for Web and mobile platforms, this change doesn't affect desktop
1. Change local time to be 30 seconds behind
2. Login to the app
3. Go to some report
4. Try to type 2-3 messages in a row
5. Messages must be in correct order

- [x] Verify that no errors appear in the JS console

### Offline tests

N/A automated tests

### QA Steps

Same as tests

- [X] Verify that no errors appear in the JS console

### PR Author Checklist

- [X] I linked the correct issue in the `### Fixed Issues` section above
- [X] I wrote clear testing steps that cover the changes made in this PR
    - [X] I added steps for local testing in the `Tests` section
    - [X] I added steps for the expected offline behavior in the `Offline steps` section
    - [X] I added steps for Staging and/or Production testing in the `QA steps` section
    - [X] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [X] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [X] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [X] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [X] I ran the tests on **all platforms** & verified they passed on:
    - [X] Android / native
    - [X] Android / Chrome
    - [X] iOS / native
    - [x] iOS / Safari
    - [X] MacOS / Chrome / Safari
    - [X] MacOS / Desktop
- [X] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [X] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [X] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [X] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [X] I verified that comments were added to code that is not self explanatory
    - [X] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [X] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [X] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [X] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [X] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [X] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [X] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [X] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [X] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [X] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [X] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [X] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [X] I verified that if a function's arguments changed that all usages have also been updated correctly
- [X] If a new component is created I verified that:
    - [X] A similar component doesn't exist in the codebase
    - [X] All props are defined accurately and each prop has a `/** comment above it */`
    - [X] The file is named correctly
    - [X] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [X] The only data being stored in the state is data necessary for rendering and nothing else
    - [X] If we are not using the full Onyx data that we loaded, I've added the proper selector in order to ensure the component only re-renders when the data it is using changes
    - [X] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [X] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [X] All JSX used for rendering exists in the render method
    - [X] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [X] If any new file was added I verified that:
    - [X] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [X] If a new CSS style is added I verified that:
    - [X] A similar style doesn't already exist
    - [X] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [X] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [X] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [X] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [X] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [X] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [X] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos

N/A automated tests
<details>
<summary>Web</summary>

https://github.com/Expensify/App/assets/39777589/538b2c62-4b63-41ee-81f0-adbb89e842d1

</details>

<details>
<summary>Mobile Web</summary>

https://github.com/Expensify/App/assets/39777589/332489b4-357e-44e1-8aab-ca38e3aa0ddd

</details>

<details>
<summary>Desktop</summary>

https://github.com/Expensify/App/assets/39777589/a7dffa6b-4c1c-4f4b-a48a-11c92bfd1ca8

</details>

<details>
<summary>iOS</summary>

https://github.com/Expensify/App/assets/39777589/cd299ab8-d7f1-4879-8b54-124be36da9f9

</details>

<details>
<summary>Android</summary>

</details>

### CODE REVIEW

**Comment by situchan** [comment]
I applied changes from this PR locally. I will see if there's any unexpected behavior for next few days.

**Comment by situchan** [comment]
Also, I have ever reproduced [this weird bug](https://github.com/Expensify/App/issues/28656#issuecomment-1743998927) but not found clear repro step yet.


Screenshot 2023-11-02 at 5 17 09 PM

**Comment by situchan** [comment]
So far I am not able to reproduce any of issues above.
@waterim if same to you, I will ask internally to generate QR build and perform regression test.
Please let me know.

**Comment by waterim** [comment]
I tried, but Im not able to reproduce them.. :(

**Comment by situchan** [comment]
@mountiny can you please generate QR build?

**Comment by situchan** [comment]
@waterim please pull main

**Comment by mountiny** [comment]
running

**Comment by situchan** [comment]
@mountiny can you please test this branch and confirm https://github.com/Expensify/App/issues/28649 is still reproducible? As it was reported by you.

**Comment by mountiny** [comment]
Unfortunately I am still seeing delays
https://github.com/Expensify/App/assets/36083550/fcf1db94-97d3-4162-a95a-d95efed039d5

**Comment by waterim** [comment]
Super weird

**Comment by situchan** [comment]
Might be related to local time. @mountiny are you currently synced with internet time? (no 1ms diff)

**Comment by mountiny** [comment]
> are you currently synced with internet time? (no 1ms diff)

Idk but I dont think this should be something we should expect from our customers, we should accommodate for that

**Comment by waterim** [comment]
After merging main to this PR Im getting this error when trying to "npm run web' 
image
@situchan @mountiny  Do you know how to fix it? (same on main as well)

**Comment by situchan** [comment]
@waterim no issue on `npm run web`. Did you `npm i`?

**Comment by waterim** [comment]
yes, did install, tried on older branches, works fine, but main doesnt work for some reason, will try to remove node.modules

**Comment by waterim** [comment]
@mountiny Can you please try to redeploy this PR?
(locally we can't reproduce this issue)

**Comment by mountiny** [comment]
Triggered again

**Comment by waterim** [comment]
Wow, it fixed  :D
@mountiny @situchan

**Comment by situchan** [comment]
> Wow, it fixed :D @mountiny @situchan

@waterim so you reproduced before https://github.com/Expensify/App/pull/30605/commits/fbe9c2836315fc74b4170c345c32efe66a8d03bd and not reproduced after https://github.com/Expensify/App/pull/30605/commits/fbe9c2836315fc74b4170c345c32efe66a8d03bd?

**Comment by waterim** [comment]
@situchan I was able to reproduce on the first PR deploy and after my fix I can't reproduce it on the deploy.
@mountiny Can you please check it too?

**Comment by waterim** [comment]
@mountiny @situchan 
Friendly bump

**Comment by situchan** [comment]
Still not able to reproduce issues mentioned in https://github.com/Expensify/App/pull/30605#issuecomment-1790532477.
But still we cannot merge this without confirming fixes.
Shall we ask QA team to do regression test?

**Comment by waterim** [comment]
@mountiny Do you able to reproduce this one?
After my fix I can't reproduce it on the newest PR deploy

**Comment by mountiny** [comment]
I cannot repro now

@waterim can you merge in main? seems like the reassure tests are failing

**Comment by waterim** [comment]
@mountiny Nice to hear!
Out of office now, will merge in 1-2 hours, thank you!

**Comment by situchan** [comment]
We're not happy yet as still [another 2 potential regressions](https://github.com/Expensify/App/pull/30605#issuecomment-1790532477) which were not reproducible by us

**Comment by waterim** [comment]
@mountiny Merged

**Comment by mountiny** [comment]
Asked QA to test

**Comment by mountiny** [comment]
they could not reproduce so I think we are good to go then!

File: src/libs/HttpUtils.js
Comment (mountiny) [inline src/libs/HttpUtils.js:34]: Lets use a bit more specific name for this regex

Comment (mountiny) [inline src/libs/HttpUtils.js:34]:
```suggestion
const APICommandRegex = /[?&]command=([^&]+)/;
```

**Review by mountiny — DISMISSED** [review]
one NAB @situchan could you please complete the checklist

**Comment by mountiny** [comment]
@waterim @situchan could you check this one out if its specific for this PR, feel like it should not

**Comment by mountiny** [comment]
@waterim @situchan could you check this one out if its specific for this PR, feel like it should nothttps://github.com/Expensify/App/issues/28656#issuecomment-1802273863

**Comment by situchan** [comment]
I am not able to reproduce on this branch merged with latest main locally

https://github.com/Expensify/App/assets/108292595/4431a9ef-e822-411c-9e91-e4c93a3cd5e4

**Comment by situchan** [comment]
This is expected, right? I adjusted time 1 hr behind


https://github.com/Expensify/App/assets/108292595/953e4104-27c8-4a96-88c4-b329c0081cf2

**Comment by situchan** [comment]
@waterim bump on https://github.com/Expensify/App/pull/30605#issuecomment-1802307261.
So what I am concerned is app should be aware of dynamic time change without refresh

**Comment by situchan** [comment]
> @waterim bump on [#30605 (comment)](https://github.com/Expensify/App/pull/30605#issuecomment-1802307261). So what I am concerned is app should be aware of dynamic time change without refresh.


So you meant no need to care about this, right?

**Comment by situchan** [comment]
Let's fix job1 failure. And will do checklist

**Comment by waterim** [comment]
@situchan I think so, this solution was an idea from David Barett

**Comment by waterim** [comment]
@situchan fixed

**Comment by waterim** [comment]
@situchan friendly bump :)

**Comment by situchan** [comment]
@waterim I managed reproduce on desktop app. Other platforms work fine.


https://github.com/Expensify/App/assets/108292595/61f29dab-7054-4bea-a454-2f4e25ade6fe


https://github.com/Expensify/App/assets/108292595/e4ba4472-033a-41f9-9461-e6db6a2027a3

**Comment by waterim** [comment]
Hello guys, I will be OOO on Monday and Thuesday. 
Will continue working on Wednesday!

**Comment by waterim** [comment]
@situchan I found an issue
Web is returning those headers:
image
While desktop only these:
image
And it doesnt have any date inside

**Comment by waterim** [comment]
Okay, yes, the main issue here is that electron fetch probably is not returning all headers and thats why I don't have an access to them at all :(

**Comment by waterim** [comment]
Can we try to test it on a deploy? @situchan

**Comment by situchan** [comment]
> Can we try to test it on a deploy? @situchan

It header isn't returning date, at least it should not break anything on main when user local time is correct

**Comment by waterim** [comment]
@situchan Yes, I added a verification to check if it exists or not

**Comment by situchan** [comment]
@waterim did you also test on QR release build?

**Comment by mountiny** [comment]
I am running a test build where we can see it on desktop hopefully, the bug is not really acceptable at least from what I can see from the video

**Comment by waterim** [comment]
@mountiny Access denied for desktop one :(

**Comment by mountiny** [comment]
@waterim we are expereincing issue with the build, I think we can wait for that to be resolved

**Comment by waterim** [comment]
@situchan @mountiny Pushed latest changes

**Comment by mountiny** [comment]
Running the build, I am still kinda confused if we need this change

**Comment by waterim** [comment]
Okay, looks like delay appears on the desktop app with changed time, but works fine with default auto-time. With all other platforms it works good. 
Don't know exactly why electron fetch doesn't have all headers to be honest.

What do you think guys regarding this?
The only issue is on desktop with disabled auto datetime, thats 0.0001% of users

**Comment by situchan** [comment]
> The only issue is on desktop with disabled auto datetime, thats 0.0001% of users

Also, this is not a regression but production behavior. So desktop isn't affected by this PR.

In summary:
desktop: doesn't fix https://github.com/Expensify/App/issues/21490 but same as production
web: fix https://github.com/Expensify/App/issues/21490

As long as this PR doesn't cause regression, I am fine to proceed. Also there's possibility that production desktop app might work fine.

**Comment by mountiny** [comment]
@situchan to confirm, no new regression, issue fixed on web and whatever the issue is on desktop that is already in production?

If so could you approve this PR please?

@waterim could you update the test steps to clearly state what platform should be tested for this?

**Review by situchan — APPROVED** [review]
> @situchan to confirm, no new regression, issue fixed on web and whatever the issue is on desktop that is already in production?

right

> If so could you approve this PR please?

**Comment by mountiny** [comment]
Oh wow I am not even assigned to the linked issue 😂 

look at that

**Comment by situchan** [comment]
## Reviewer Checklist

- [x] I have verified the author checklist is complete (all boxes are checked off).
- [x] I verified the correct issue is linked in the `### Fixed Issues` section above
- [x] I verified testing steps are clear and they cover the changes made in this PR
    - [x] I verified the steps for local testing are in the `Tests` section
    - [x] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [x] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I verified tests pass on **all platforms** & I tested again on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [x] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>



</details>

<details>
<summary>Android: mWeb Chrome</summary>



</details>

<details>
<summary>iOS: Native</summary>



</details>

<details>
<summary>iOS: mWeb Safari</summary>



</details>

<details>
<summary>MacOS: Chrome / Safari</summary>



</details>

<details>
<summary>MacOS: Desktop</summary>



</details>

File: src/libs/HttpUtils.ts
Comment (mountiny) [inline src/libs/HttpUtils.ts:62]: I think it would be great to leave a comment @waterim about why we are even doing this

**Review by mountiny — CHANGES_REQUESTED** [review]

**Comment by situchan** [comment]
> Oh wow I am not even assigned to the linked issue 😂
> 
> look at that

Ah, the reason why I tagged you first time is that you reported https://github.com/Expensify/App/issues/28649 which was not reproducible by us

**Comment by waterim** [comment]
Updated everything @mountiny

**Review by mountiny — APPROVED** [review]
Alright, thanks! All yours @nkuoch

**Review by nkuoch — APPROVED** [review]

**Pull request merged**
